### ISSUE TITLE
Mediainfo return runtime error when fetching info from S3 Pre-signed URL

### ISSUE THREAD

**Issue by changma2** [MediaArea/MediaInfoLib#2015]
Hi there, we are experiencing runtime error when trying to get Media info for some production files. 
Sorry I could not share the files or the links. 
To give more background info, we are using v23.07 in our service, and we've already done some debug processes, 1. we downloaded the files to our local, and run mediainfo(v24.01) on them and all the files succeeded, 2. in order to make sure it's not the version problem, we create a S3 signed-url and try to use it in the local mediainfo, but it's stuck at running state with no output, need to ctrl+c to exit.
Please help on this, any help is appreciated, thank you.

**Comment by JeromeMartinez** [comment]
> any help is appreciated

Our CI still works well with S3 files, so we can not reproduce and can not help without a link.

> Sorry I could not share the files or the links.

Without that, it will be difficult to help in remote, longer for us and not part of free support (we have [professional services](https://mediaarea.net/Services) if you need a SLA and/or prioritizing this issue), we would have to run a several sdebug version logging all what is done up to the crash, having a file to be tested on our side with our debug environment would be a lot quicker.
Please send a private link to info@mediaarea.net. We can also sign a NDA if needed (and we don't play the file).

**Comment by changma2** [comment]
@JeromeMartinez After talking to our security team, we can share content with signed NDA and have a proof of deletion after debugging on the issues. We aggregate all the issues we've seen so far and we can share a checklist along with the contents. Regarding the professional services, we think it's suitable for our currently use case, but we will need more discussion and decision making before we can move forward on that. 
Moreover, since most of the issues here are runtime error when using pre-signed url, how do you want me to share the contents, do you want me to share a list of 12-hr duration pre-signed url? Or you have a way to created signed-url for the files, so I can share BOX link of the original files?
Lastly, for the email `info@mediaarea.net`, our security team want me to confirm if this is a shared address or it's an individual user, if it's a shared address, can you please provide an individual user address? 
Thanks for your help.

**Comment by changma2** [comment]
@JeromeMartinez NDA has been sent to `info@mediaarea.net`, please check and let me know if you having trouble finding or accessing it. Thank you.

**Comment by JeromeMartinez** [comment]
> Or you have a way to created signed-url for the files, so I can share BOX link of the original files?

We could first try to upload them on our S3 and create the signed URLs ourselves, and we'll let you know if we can't reproduce the issues.

> Lastly, for the email info@mediaarea.net, our security team want me to confirm if this is a shared address or it's an individual user, 

Currently alias of my email address (the one used for commits on this repo), as the manager of the company.

> @JeromeMartinez NDA has been sent to info@mediaarea.net, please check and let me know if you having trouble finding or accessing it. Thank you.

Received.

**Comment by changma2** [comment]
@JeromeMartinez I will share the files to the address with your alias shortly.

**Comment by changma2** [comment]
@JeromeMartinez We created a Box folder and shared with the address with your alias, please let us know once you downloaded all the files, so we can retrieve access. Thank you!

**Comment by JeromeMartinez** [comment]
> please let us know once you downloaded all the files

Done.

Remind that as we have no service contract so the response time is not guarantied, but we'll have a look on it this week (we dislike having crashes/freezes in production wherever it is).

**Comment by changma2** [comment]
> Done.

Thank you for your update!

> Remind that as we have no service contract so the response time is not guarantied, but we'll have a look on it this week (we dislike having crashes/freezes in production wherever it is).

Understood, and appreciate your help and support again.

**Comment by changma2** [comment]
@JeromeMartinez Hi there, thanks for actively working on this issue and great to see you closed it, but before I let you go, wondering since we are using mediainfo as a layer in lambda, can I just pull the latest change and build a layer from source and it should work right?

**Comment by JeromeMartinez** [comment]
@changma2 we succeeded to reproduce the issue (infinite loop) for most of the files but not all, anyway we expect to have fixed all of the issues with an infinite loop / stall so we bet that all stall issues are fixed even on your side.
Please test the [development snapshots](https://mediaarea.net/download/snapshots/binary/) and confirm.

By the way, we also worked on reducing the amount of content read from the file before giving up when the format is not detected (i.e. with your files, as they are not multimedia content), it is mainly for catching raw streams (raw AVC, raw HEVC, etc) not starting at the beginning of the file, by default it reads the first 16 MiB and you can, if you don't have such raw content) with an option e.g. " --FormatDetection_MaximumOffset=1000000" but FYI it is not intensively tested.

We did that within free support because we dislike a lot crashes and/or stall and it could happen for other users/customers so we prefer to fix it before they face that in production but please consider our support contract for a long term relationship in case you need more fixes and/or support of more formats.

Note that we started MediaInfo(Lib) with local storage in mind, having disk cache, so we didn't (yet) optimize the count of seeks or the size of the content we read, and we are looking for sponsors for that ;-) if you are interested in reducing the I/O used by MediaInfo.

About the CPL file, we see no issue but we have only the master file without the referenced files, so impossible to see what you consider wrong with it, you may need to share the other files. But it may be not enough, as the CPL has only uuid of the files, if you want us to analyze the other files from the CPL file, we'll need to implement a way to map the uuids to a URL, please open a new ticket dedicated to that if you are interested in such feature.
Another possibility, in parallel, is to report A/V metadata based only on CPL content, currently not implemented but your CPL has a lot of information we could report e.g. video height or audio language, and we could also do some [conformance checking](https://mediaarea.net/MediaConch) (the website is a bit outdated, now with have [more formats supported](https://mediaarea.net/Specs), e.g. we have [now a full ADM conformance checker](https://www.linkedin.com/feed/update/urn:li:activity:7158512646235865088) as well as a full USAC/xHE-AAC conformance checker) e.g. checking the coherence of the XML file as well as the coherence between the metadata in the CPL and the metadata in MXF files.

**Comment by JeromeMartinez** [comment]
> can I just pull the latest change and build a layer from source and it should work right?

Right. Either from source or [from our development snapshots](https://mediaarea.net/download/snapshots/binary/).
We don't know much about Lambda but maybe our Lambda specific builds ([x86_64](https://mediaarea.net/download/snapshots/binary/mediainfo/20240502/MediaInfo_CLI_24.04.20240502_Lambda_x86_64.zip), [arm64](https://mediaarea.net/download/snapshots/binary/mediainfo/20240502/MediaInfo_CLI_24.04.20240502_Lambda_arm64.zip)) work out of the box with your use case.

**Comment by changma2** [comment]
@JeromeMartinez For the CPL file, what we consider wrong is, it's actually an xml file, so we are expecting to only have a type:text from the output, but instead, we are seeing videocount:7, audiocount:7, textcount:7 from the response. We know the file contains those metadata, but xml is not an extension for video and audio. That's the confusion part for the CPL file.

**Comment by changma2** [comment]
I'm building the lambda layer right now and will re-drive all the failed files except the CPL one and I will report back here.

**Comment by changma2** [comment]
@JeromeMartinez Confirmed all the files except the CPL file are working properly now, thank you again for your support and we definitely going to consider and discuss about the professional support plan you guys provided.

**Comment by JeromeMartinez** [comment]
>  it's actually an xml file,

the .cpl file is a XML file as much as a MP4 file is an octet stream: this is what you see when the A/V format is not supported.

> so we are expecting to only have a type:text from the output

The A/V content of the CPL is **not** only a text file, it contains 7 video, 7 audio, 7 text.

> but xml is not an extension for video and audio. 

Here the extension is .cpl, not .xml, but actually an extension is a convention and not mandatory so 2 issues: 1/ your file has no xml extension 2/ your file contains A/V content as defined by CPL format.
Here XML is just an intermediate technical layer for storing A/V content.
Like a xlsx (Excel file format) uses ZIP as an intermediate layer, xlsx is a ZIP file for a pure neutral ZIP decompressor and an Excel file for a sheet processor, here a CPL file is a XML file for a pure neutral XML reader and a A/V content for a video player.

> That's the confusion part for the CPL file.

What you don't get is that the CPL file is exactly as a MP4 file: a bunch of bytes storing some A/V content, here the CPL file is storing A/V content (in practice a bunch of metadata + link to other files, but for a A/V player it is transparent, it is some A/V content to display; and some MP4 file do exactly same, no actual A/V content and just links to other files, this XML file and some MP4 files have exactly the same structure without any actual data but links to actual data).

CPL is standardized by the SMPTE ("Society of **Motion Picture** and Television Engineers"), it is A/V content and MediaInfo analyzes A/V content so it is the purpose of MediaInfo to display that your CPL file has 7 video, 7 audio, 7 text, it is what is in it and what would be managed by a CPL aware A/V player..
If you want to explicitly exclude some A/V formats you don't like from the analysis, it would possible with an option (not yet implemented).

TLDR: it is the purpose of MediaInfo to analyze A/V content including CPL format and the current MediaInfo behavior is the expected one.

**Comment by changma2** [comment]
@JeromeMartinez Wow, that's actually great explanation. We will discuss as a team on how to dealing with file like this one. Thank you again.
